// @fast-crud/fast-extends/dist/uploader-form-42acdd99.mjs
import l from "lodash-es";
import { b as d, u as f } from "./index-d86da30d.mjs";
import "vue";
import "@fast-crud/fast-crud";
function c(e, n, s) {
  let t;
  s.response ? t = `${s.response.error || s.response}` : s.responseText ? t = `${s.responseText}` : t = `fail to post ${e} ${s.status}`;
  const o = new Error(t);
  return o.status = s.status, o.method = "post", o.url = e, o;
}
function p(e) {
  const n = e.responseText || e.response;
  if (!n)
    return n;
  try {
    return JSON.parse(n);
  } catch {
    return n;
  }
}
function m(e, n, s) {
  if (typeof XMLHttpRequest > "u")
    return;
  const t = new XMLHttpRequest(), o = e.action;
  t.timeout = e.timeout, t.upload && (t.upload.onprogress = function(r) {
    r.total > 0 && (r.percent = r.loaded / r.total * 100), e.onProgress(r);
  });
  const i = new FormData();
  e.data && Object.keys(e.data).forEach((a) => {
    i.append(a, e.data[a]);
  }), i.append(e.name, e.file, e.file.name), t.onerror = function(r) {
    s(r);
  }, t.onload = function() {
    if (t.status < 200 || t.status >= 300)
      return e.onError(c(o, e, t));
    n(p(t));
  }, t.open("post", o, !0), e.withCredentials && "withCredentials" in t && (t.withCredentials = !0);
  const u = e.headers || {};
  for (const a in u)
    u.hasOwnProperty(a) && u[a] !== null && t.setRequestHeader(a, u[a]);
  return t.send(i), t;
}
function y(e) {
  return new Promise((n, s) => {
    m(e, async (t) => {
      n(t);
    }, (t) => {
      s(t);
    });
  });
}
async function w(e) {
  const { file: n, fileName: s, onProgress: t } = e, o = e.options, i = await d(n, s, o);
  o.data == null && (o.data = {}), o.data.key = i;
  const u = {
    file: n,
    onProgress: t,
    timeout: 6e4,
    ...o
  };
  delete u.uploadRequest;
  let r = await (o.uploadRequest ?? y)(u);
  return o.successHandle && (r = await o.successHandle(r, u)), r && typeof r == "object" && r.key == null && (r.key = i), r;
}
async function H(e) {
  const { getConfig: n } = f(), s = n("form");
  return e.options = l.merge({}, l.cloneDeep(s), e.options), await w(e);
}
export {
  H as upload
};
