// @fast-crud/ui-interface/dist/ui-interface.mjs
import { ref as m, resolveComponent as g, createVNode as C, computed as h, defineComponent as U } from "vue";
import i from "lodash-es";
class x {
  constructor() {
    this.ref = m(null);
  }
  set(o) {
    this.ref.value = o;
  }
  get() {
    if (this.ref.value == null)
      throw new Error("您还未设置ui,先安装依赖@fast-crud/ui-interface,然后在use(FastCrud)前安装ui，app.use(UiXxx)");
    return this.ref.value;
  }
}
const d = new x();
function R() {
  return {
    uiContext: d,
    ui: d.get(),
    set: (r) => {
      d.set(r);
    }
  };
}
const l = (r) => {
  const o = typeof r.is == "string" ? g(r.is) : r.is;
  return C(o, r.props, r.slots);
}, c = (r, o) => l(r.builder(o)), p = (r, o, n) => {
  const t = {}, s = r.modelValue, e = o.vModel;
  s != null && e && (e != null && e.get && (e != null && e.set) ? (t[s] = e.get(), t[`onUpdate:${s}`] = (u) => {
    e.set(u), e.onChange && e.onChange(u);
  }) : e.ref && e.key ? (t[s] = i.get(e.ref, e.key), t[`onUpdate:${s}`] = (u) => {
    i.set(e.ref, e.key, u), e.onChange && e.onChange(u);
  }) : console.warn("vModel配置错误:", r, e));
  const a = {
    is: o.is || r.name,
    props: o.props,
    slots: o.slots
  }, f = n instanceof Function ? n() : n;
  return i.merge({
    props: t
  }, f, a);
}, F = (r, o = {}) => {
  const n = {
    ...r
  };
  return n.render = (t) => c(n, t), n.builder || (n.builder = (t) => p(n, t, o)), n.buildProps = (t) => n.builder(t).props, n.builderComputed = (t) => h(() => n.builder(t)), n;
};
function k() {
  return {
    creator: F,
    doRenderComponent: l,
    renderComponent: c,
    buildBinding: p
  };
}
const B = /* @__PURE__ */ U({
  name: "FsUiRender",
  props: {
    renderFn: {
      type: Function,
      default() {
        return () => null;
      }
    }
  },
  setup(r) {
    return () => r.renderFn();
  }
});
export {
  x as UiContext,
  B as UiRender,
  p as buildBinding,
  F as creator,
  l as doRenderComponent,
  c as renderComponent,
  d as uiContext,
  R as useUi,
  k as useUiRender
};
