// @fast-crud/fast-crud/dist/fast-crud.mjs
import { A as l, ao as u, C as i, D as n, F as m, a7 as c, ak as p, O as C, $ as d, a9 as b, L as D, I as g, Y as f, aj as T, ag as x, af as y, ah as S, ac as h, ad as L, ai as P, al as E, a1 as I, a0 as R, U as O, W as V, V as v, _ as A, X as M, Q as U, S as k, R as w, T as B, K as H, J as W, P as j, H as q, N as G, a2 as K, a3 as N, a5 as z, a4 as J, M as Q, Z as X, ab as Y, am as Z, a6 as _, a8 as $, ae as ss, G as as, an as es, i as os, q as ts, h as rs, z as Fs, F as ls, l as us, ar as is, v as ns, t as ms, aa as cs, E as ps, as as Cs, ap as ds, aq as bs, B as Ds, r as gs, s as fs, w as Ts, j as xs, b as ys, k as Ss, m as hs, x as Ls, o as Ps, y as Es, d as Is, e as Rs, f as Os, a as Vs, p as vs, n as As, u as Ms } from "./index-0bad1a5e.mjs";
export * from "@fast-crud/ui-interface";
import { uiContext as ws, useUi as Bs } from "@fast-crud/ui-interface";
import "vue";
import "lodash-es";
import "dayjs";
import "@iconify/vue";
import "vue-router";
export {
  l as AsyncComputeValue,
  u as ColumnsFilterProvideKey,
  i as ComputeValue,
  n as Dict,
  m as FastCrud,
  c as FsActionbar,
  p as FsBox,
  C as FsButton,
  d as FsCell,
  b as FsColumnsFilterLayoutDefault,
  D as FsComponentRender,
  g as FsContainer,
  f as FsCrud,
  T as FsDateFormat,
  x as FsDictCascader,
  y as FsDictCascaderFormat,
  S as FsDictCheckbox,
  h as FsDictRadio,
  L as FsDictSelect,
  P as FsDictSwitch,
  E as FsDictTree,
  I as FsEditable,
  R as FsEditableCell,
  O as FsForm,
  V as FsFormHelper,
  v as FsFormItem,
  A as FsFormProvider,
  M as FsFormWrapper,
  U as FsIcon,
  k as FsIconSvg,
  w as FsIconify,
  B as FsLabel,
  H as FsLayoutCard,
  W as FsLayoutDefault,
  j as FsLoading,
  q as FsPage,
  G as FsRender,
  K as FsRowHandle,
  N as FsSearch,
  z as FsSearchLayoutDefault,
  J as FsSearchV1,
  Q as FsSlotRender,
  X as FsTable,
  Y as FsTableColumnsFixedController,
  Z as FsTableSelect,
  _ as FsTabsFilter,
  $ as FsToolbar,
  ss as FsValuesFormat,
  as as GlobalConfig,
  es as SetFormDataOptions,
  os as asyncCompute,
  ts as buildTableColumnsFlatMap,
  rs as compute,
  Fs as crudOptionsPlugins,
  ls as default,
  us as dict,
  is as exportTable,
  ns as forEachColumns,
  ms as forEachTableColumns,
  cs as fsColumnsFilterNestList,
  ps as getCrudOptionsPlugin,
  Cs as importTable,
  ds as loadFsExportUtil,
  bs as loadFsImportUtil,
  Ds as registerCrudOptionsPlugin,
  gs as registerMergeColumnPlugin,
  fs as setLogger,
  ws as uiContext,
  Ts as useColumns,
  xs as useCompute,
  ys as useCrud,
  Ss as useDict,
  hs as useDictDefine,
  Ls as useDrag,
  Ps as useExpose,
  Es as useFormWrapper,
  Is as useFs,
  Rs as useFsAsync,
  Os as useFsRef,
  Vs as useI18n,
  vs as useMerge,
  As as useTypes,
  Bs as useUi,
  Ms as utils
};
